contracts-periphery/src/UmbraBatchSend.sol

```diff
@@ -24,69 +27,133 @@ contract UmbraBatchSend {
     bytes32 ciphertext;
   }
 
+  mapping(address => uint256) internal TotalTransferAmountPerToken;
+  struct TransferSummary {
+    uint256 amount;
+    address tokenAddr;
+  }
+
   error ValueMismatch();
+  error TransferAmmountMismatch();
   event BatchSendExecuted(address indexed sender);
 
   constructor(IUmbra _umbra) {
     umbra = _umbra;
   }
 
   function batchSendEth(uint256 _tollCommitment, SendEth[] calldata _params) external payable {
-    if(msg.value != _sumValueFromEthSends(_tollCommitment, _params)) revert ValueMismatch();
+    if (msg.value != _sumValueFromEthSends(_tollCommitment, _params)) revert ValueMismatch();
     _batchSendEth(_tollCommitment, _params);
     emit BatchSendExecuted(msg.sender);
   }
 
-  function batchSendTokens(uint256 _tollCommitment, SendToken[] calldata _params) external payable {
-    if(msg.value != _tollCommitment * _params.length) revert ValueMismatch();
-    _batchSendTokens(_tollCommitment, _params);
+  function batchSendTokens(
+    uint256 _tollCommitment,
+    SendToken[] calldata _params,
+    TransferSummary[] calldata _transferSummary
+  ) public payable {
+    if (msg.value < _tollCommitment * _params.length) revert ValueMismatch();
+
+    for (uint256 i = 0; i < _params.length; i++) {
+      // Used later to validate total amounts are correctly provided
+      TotalTransferAmountPerToken[_params[i].tokenAddr] += _params[i].amount;
+    }
+
+    uint256 summaryLength = _transferSummary.length;
+    for (uint256 j = 0; j < summaryLength; ) {
+      IERC20 token = IERC20(address(_transferSummary[j].tokenAddr));
+
+      if (TotalTransferAmountPerToken[_transferSummary[j].tokenAddr] != _transferSummary[j].amount)
+        revert TransferAmmountMismatch();
```

Comment:
nit: if have a newline, I think we should include the braces

Reply:
Yup, I confused myself for a sec there too https://github.com/ScopeLift/umbra-protocol/blob/bf77a1765f7683b3be691c8e28a230a0711c9279/contracts-periphery/src/UmbraBatchSend.sol#L66-L68